lib/babe/parachain_inherents.go

```diff
@@ -0,0 +1,345 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package babe
+
+import (
+	"fmt"
+
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+// signature could be one of Ed25519 signature, Sr25519 signature or ECDSA/SECP256k1 signature.
+type signature [64]byte
+
+// validityAttestation is an implicit or explicit attestation to the validity of a parachain
+// candidate.
+type validityAttestation scale.VaryingDataType
+
+// Set will set a VaryingDataTypeValue using the underlying VaryingDataType
+func (va *validityAttestation) Set(val scale.VaryingDataTypeValue) (err error) {
+	// cast to VaryingDataType to use VaryingDataType.Set method
+	vdt := scale.VaryingDataType(*va)
+	err = vdt.Set(val)
+	if err != nil {
+		return fmt.Errorf("setting value to varying data type: %w", err)
+	}
+	// store original ParentVDT with VaryingDataType that has been set
+	*va = validityAttestation(vdt)
+	return nil
+}
+
+// Value returns the value from the underlying VaryingDataType
+func (va *validityAttestation) Value() (val scale.VaryingDataTypeValue) {
+	vdt := scale.VaryingDataType(*va)
+	return vdt.Value()
+}
+
+// implicit is for implicit attestation.
+type implicit validatorSignature
+
+// Index returns VDT index
+func (implicit) Index() uint {
+	return 1
+}
+
+// explicit is for explicit attestation.
+type explicit validatorSignature
+
+// Index returns VDT index
+func (explicit) Index() uint {
+	return 2
+}
+
+// newValidityAttestation creates a ValidityAttestation varying data type.
+func newValidityAttestation() validityAttestation {
+	vdt, err := scale.NewVaryingDataType(implicit{}, explicit{})
+	if err != nil {
+		panic(err)
+	}
+
+	return validityAttestation(vdt)
+}
+
+// disputeStatement is a statement about a candidate, to be used within the dispute
+// resolution process. Statements are either in favour of the candidate's validity
+// or against it.
+type disputeStatement scale.VaryingDataType
+
+// Set will set a VaryingDataTypeValue using the underlying VaryingDataType
+func (distputedStatement *disputeStatement) Set(val scale.VaryingDataTypeValue) (err error) {
+	// cast to VaryingDataType to use VaryingDataType.Set method
+	vdt := scale.VaryingDataType(*distputedStatement)
+	err = vdt.Set(val)
+	if err != nil {
+		return fmt.Errorf("setting value to varying data type: %w", err)
+	}
+	// store original ParentVDT with VaryingDataType that has been set
+	*distputedStatement = disputeStatement(vdt)
+	return nil
+}
+
+// Value will return value from underying VaryingDataType
+func (distputedStatement *disputeStatement) Value() (val scale.VaryingDataTypeValue) {
```

Comment:
Typo `distputedStatement`, at least rename to `disputedStatement`. Ideally rename to `d` to avoid any confusion (past dispute or present dispute?)